### PR TITLE
Mielensab

### PR DESCRIPTION
**Pull request by briandleahy**
This extends `holopy` with the ability to calculate the image of single spheres through a lens with spherical aberration of arbitrary order, at no extra computational cost compared to that for an unaberrated lens (and therefore still faster than the calculations without a lens).

@carolinesmartin has been using this for about 1 year now, so the code has been pretty well stress-tested. 

@barkls let me know if this should be added to a user guide.

### CODE REVIEW

File: holopy/scattering/theory/mielens.py [hunk 1]
```diff
@@ -128,3 +131,33 @@ def raw_fields(self, positions, scatterer, medium_wavevec, medium_index,
...
+        return field_calculator
+
+
+class AberratedMieLens(MieLens):
+    def __init__(self, spherical_aberration=0.0, lens_angle=1.0,
+                 calculator_accuracy_kwargs={}):
+        super(AberratedMieLens, self).__init__()
+        self.lens_angle = lens_angle
+        self.spherical_aberration = spherical_aberration
+        self.calculator_accuracy_kwargs = calculator_accuracy_kwargs
```
Comment (barkls) [inline holopy/scattering/theory/mielens.py:151]: Two of these lines (`lens_angle` and `calculator_accuracy_kwargs`) are sort of redundant with the call to `super.__init__()`, except that you don't pass them through to the `MieLens` init. Is there a specific reason you chose to do things this way?

File: holopy/scattering/theory/mielensfunctions.py [hunk 2]
```diff
@@ -264,11 +264,60 @@ def _interpolate_and_eval_mielens_i_n(self, krho, n=0):
...
+        return self.particle_kz * (1 - self._quad_pts)
+
     def _check_parameters(self):
-        must_be_specified = ['particle_kz', 'index_ratio',
-                             'size_parameter', 'lens_angle']
-        if any([getattr(self, p) is None for p in must_be_specified]):
-            raise ValueError("{} must be specified.".format(must_be_specified))
+        if any([getattr(self, p) is None for p in self.must_be_specified]):
+            msg = "{} must be specified.".format(self.must_be_specified)
+            raise ValueError(msg)
```
Comment (barkls) [inline holopy/scattering/theory/mielensfunctions.py:273]: HoloPy actually has a `MissingParameter` exception class you might want to use here.

**Review by barkls — COMMENTED** [review]
Nice improvements to code structure and thorough testing! I just had a couple minor comments that you should feel free to ignore.

This should be added to the [Scattering Theories User Guide](https://holopy.readthedocs.io/en/master/users/theories.html) that you wrote, but we can hold off until it's published if you prefer.

File: holopy/scattering/theory/mielensfunctions.py (hunk 2 above)
Comment (briandleahy) [reply holopy/scattering/theory/mielensfunctions.py:273]: Fixed by 08b8338d

**Comment by briandleahy** [comment]
Your comments are addressed in the latest commits. I added a description of the AberratedMieLens theory in fedde9a0. I would appreciate it if you could take a quick look at that and tell me if my documentation is intelligible. Also, in the process I noticed that there were some out-of-date docs; I fixed these in baedc1ff. 

Let me know if you have other comments!

**Comment by barkls** [comment]
Good catch on updating the function names in the dev guide! Everything else looks good except the posision typo

File: docs/source/users/theories.rst [hunk 3]
```diff
@@ -96,6 +96,15 @@ Lens-Based Scattering Theories
       effects of a perfect objective lens.
     * Used for performance; ``MieLens(lens_angle)`` is much faster than calling
       ``Lens(lens_angle, Mie())`` and slightly faster than ``Mie()``.
+- :class:`.AberratedMieLens`
+    * Can handle :class:`.Sphere` objects, or :class:`.Spheres` through
+      superposision.
```
Comment (barkls) [inline docs/source/users/theories.rst:101]: Typo

**Review by barkls — APPROVED** [review]

**Pull request merged**